EZP-26110: Can register custom OutputVisitor but can't set its priority

Nattfarinn opened the pull request:
**JIRA ticket related**: https://jira.ez.no/browse/EZP-26110

You can define custom output visitors and tag it (`name: ezpublish_rest.output.visitor, regexps: something`) but it won't work with current implementation unless you load your bundle before `new eZ\Bundle\EzPublishRestBundle\EzPublishRestBundle(),` (and this way sucks). It's because of order of output visitors passed to compiler pass while kernel configuration contains Pokemon "gotta catch'em all" regexp `(^.*/.*$)` so it always return `application/vnd.ez.api...` regardless of visitors and generators registered after that one.

**Solution**: Proposed solution is to implement `priority` tag attribute similar to ones used in Symfony event listeners or cache warmers with default `0` in case of missing attribute.

**Example of use**:

```yaml
custom.output.visitor.xml:
    class: '%ezpublish_rest.output.visitor.class%'
    arguments:
        - '@custom.output.generator.xml'
        - '@ezpublish_rest.output.value_object_visitor.dispatcher'
    tags:
        - { name: ezpublish_rest.output.visitor, regexps: custom.output.visitor.xml.regexps, priority: 10 }
```

bdunogier commented on eZ/Bundle/EzPublishRestBundle/Tests/DependencyInjection/Compiler/OutputVisitorPassTest.php:
```diff
@@ -53,4 +53,54 @@ public function testProcess()
+
+    public function testPriority()
+    {
+        $regexp1 = array('(^.*/.*$)');
```
The test would probably be more readable if you used one array for test data, with regexp as first element, and priority as second, optional one. You could then iterate over them to create the definitions instead of doing so sequentially.

bdunogier commented:
Looks okay to me besides the readability comment about tests. Further note about the test: would it be a good opportunity to update it to use the `AbstractCompilerPassTestCase` ?

Nattfarinn commented:
Shouldn't be a problem but I have to do some research as it's first time I see such thing :P

bdunogier commented:
It's nothing extraordinary, but really convenient. It takes care of a most of the container builder's complexity, and provides you with a set of assertions that avoid the awful method checks you have :)

Nattfarinn commented:
We won't avoid these awful assertions unfortunately, because it's the only way to check mathod calls order. `AbstractCompilerPassTestCase` gives us only `assertContainerBuilderHas...` assertions without checking call order.

Nattfarinn commented:
(...but as you suggested I used `AbstractCompilerPassTestCase` at least wherever I could and made test a little bit cleaner anyway :) )

Nattfarinn commented:
Hope to get my MacBook repaired soon or I'll kill @ezrobot, I swear.

```bash
dev@server:~/Projects/ezs-clean.dev/src/ezpublish-kernel$ php-cs-fixer fix eZ/Bundle/EzPublishRestBundle/Tests/DependencyInjection/Compiler/OutputVisitorPassTest.php
Fixed all files in 0.083 seconds, 4.000 MB memory used
```

Nothing changed? So everything is OK, right?

```bash
dev@server:~/Projects/ezs-clean.dev/src/ezpublish-kernel$ php-cs-fixer fix --config-file=./.php_cs eZ/Bundle/EzPublishRestBundle/Tests/DependencyInjection/Compiler/OutputVisitorPassTest.php
Loaded config from "./.php_cs"
   1) /home/dev/Projects/ezs-clean.dev/src/ezpublish-kernel/eZ/Bundle/EzPublishRestBundle/Tests/DependencyInjection/Compiler/OutputVisitorPassTest.php
Fixed all files in 0.047 seconds, 4.000 MB memory used
```

Wrong! For some reason my `php-cs-fixer` ignores config file if I don't specificaly tell him to load it from current directory. I hate to work on machine I know it is temporary...

andrerom commented:
@Nattfarinn Hakuna matata, don't blame @ezrobot ;)

Besides that +1
ping @yannickroger / @pspanja

pspanja commented:
+1
